tests/test_api.py

```diff
@@ -146,6 +147,49 @@ def test_read_write_read_compare(self) -> None:
 
             os.remove(path_2)
 
+    def assertEqualRecursively(self, var: Any) -> None:
+        """Recursively checks that __eq__ implementations work.
+        Initially, an instance of a top-level metadata class is passed and for
+        each of its attributes, that are objects implementing "__eq__", it's
+        verified that their "__eq__" implementations work.
+        Then assertEqualRecursively is called for each of those attributes.
+        """
+        if isinstance(var, dict):
+            for val in var.values():
+                self.assertEqualRecursively(val)
+
+        # Check if var is an object implementing __dict__ (needed for "vars()")
+        # and that is not a securesystemslib.Signature instance.
+        elif hasattr(var, "__dict__") and type(var).__name__ != "Signature":
+            for attr in vars(var).keys():
+                # If "var.attr" value is None the asserts below won't work
+                # properly. Solve this problem by giving a "" value to "attr".
+                if getattr(var, attr) is None:
+                    setattr(var, attr, "")
+
+                obj_2 = deepcopy(var)
+                self.assertEqual(var, obj_2)
+
+                # Assert that after changing obj_2 they are no longer equal.
+                setattr(obj_2, attr, None)
+                self.assertNotEqual(var, obj_2)
+
+                self.assertEqualRecursively(getattr(var, attr))
+
+    def test_md__eq__(self) -> None:
+        # Test that the "__eq__" implementations work for each of the top level
+        # metadata classes and for each of the helper classes.
+        for role in TOP_LEVEL_ROLE_NAMES:
+            md_obj = Metadata.from_file(f"{self.repo_dir}/metadata/{role}.json")
+            self.assertEqualRecursively(md_obj)
```

Comment:
The implementation of assertEqualRecursively looks pretty smart. But wouldn't it be enough to just call `assertEqual` on `md_obj` and a `deepcopy` of it? If your hierarchy of `__eq__` weren't implemented this should also fail.

And regarding the `assertNotEqual`, if you look at the coverage, you'll see that not all branches in your `__eq__` implementations are covered. And at least to me it's not immediately clear which condition in `__eq__` should match after your `setattr(obj_2, attr, None)`.

Maybe it would be more obvious and yield better coverage if you just iterate through a table of unequal  (for each condition you test in `__eq__`) object pairs for each type and call `asserEqual` on them.